vip_hci/leastsq/leastsq.py

```diff
@@ -23,9 +23,6 @@
 from ..conf.utils_conf import eval_func_tuple as EFT
 
 
-array = None
```

Comment:
What's the purpose of this change? The idea of having a global array is that there's no need to pass it in internal function calls. This is more performant when running in multiprocessing mode. Could you please do proper timings and check with both variants: with the global array and passing the big cube every time?

Reply:
I did a quick benchmark with `nproc=1`, where the new code was slightly faster than the "global" one (12s vs 12.5s). I'll look into it in more detail.

Reply:
The data cube should not be copied in memory when passed as function argument, as also shown by my benchmark: there is no significant performance difference between the two versions (this time with `nproc=None`):

```
# global:
    raw times: 11.1 8.18 8.53 7.88 7.78 7.91 8.01 7.99 8.11 8.47
    1 loops, best of 10: 7.78 sec per loop

# new:
    raw times: 11.4 7.53 7.6 8.21 8.82 7.97 8.42 7.72 7.77 7.96
    1 loops, best of 10: 7.53 sec per loop
```

As we are already passing many parameters to e.g. `_leastsq_patch`, the gain in readability is minimal. In fact, I find the code even more difficult to read when there are some global variables, as the helper functions are not encapsulated any more.
I propose to revert the change here, and to open a new pull request to remove the `global` variables found in the other submodules too (`frdiff/framediff.py`, `llsg/llsg.py`, `madi/adi_source.py`, `preproc/derotation.py`).

Reply:
see [commit 101e481](https://github.com/vortex-exoplanet/VIP/pull/149/commits/101e481c16680312080a0f2679ec544c6f9095f2)

Reply:
benchmark script:

``` python
# timing_xloci.py
# for python 3.6+

import sys, os

global MODULEDIR, TUTDIR
MODULEDIR = "/Users/ralf/.local/share/virtualenvs/pyc-JlLL1mk7/src"
TUTDIR = "/Users/ralf/Documents/KIT/UGA/MA/py/r4lv-vip-tutorial"

def test(module, nproc=1):
    sys.path.insert(0, f"{MODULEDIR}/{module}")

    import vip_hci as vip
    print("loading VIP version {} from <VENV>/{}".format(vip.__version__,
          os.path.relpath(os.path.dirname(vip.__file__), MODULEDIR)))



    cube_orig, angs = vip.fits.open_adicube(f"{TUTDIR}/naco_betapic.fits",
                                            verbose=False)

    cube = vip.preproc.cube_recenter_dft_upsampling(cube_orig, 51, 51,
                                                    fwhm=4.6,
                                                    subi_size=5, negative=True,
                                                    full_output=False,
                                                    debug=False, verbose=False)

    fr_xloci_scl2 = vip.leastsq.xloci(cube, angs, fwhm=4.6, nproc=nproc,
                                      verbose=False)


```

**usage**

First, modify the `MODULEDIR` and `TUTDIR` to point to the corresponding directories on your machine. In the `MODULEDIR`, there are 2 subdirectories which contain the entire VIP code, with corresponding changes to the implementation. The module to test can be selected by passing the folder name to the `test()` function. Run the benchmark with:

``` bash
python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci", None)'
```

and

``` bash
python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci-global", None)'
```

**results**
``` text
single processor
    old/global
        ❯ pipenv run python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci-global", 1)' 
        raw times: 16.3 12.9 13.5 13 13.1 12.7 12.8 12.8 13.4 13.1
        1 loops, best of 10: 12.7 sec per loop
    new
        ❯ pipenv run python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci", 1)'  
        raw times: 15.2 11.9 14.3 14.8 13.2 12.3 12.6 13.3 13.3 12.4
        1 loops, best of 10: 11.9 sec per loop
multi processor
    old/global
        ❯ pipenv run python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci-global", None)'
        raw times: 11.1 8.18 8.53 7.88 7.78 7.91 8.01 7.99 8.11 8.47
        1 loops, best of 10: 7.78 sec per loop
    new
        ❯ pipenv run python -m timeit -s 'import timing_xloci as t' -n1 -r10 -v 't.test("vip-hci", None)' 
        raw times: 11.4 7.53 7.6 8.21 8.82 7.97 8.42 7.72 7.77 7.96
        1 loops, best of 10: 7.53 sec per loop
```

Reply:
Thank you for running the tests. I wasn't very detailed when I mentioned that using the global array the performance was better. Two things need to be tested: memory consumption and  running time. There are a couple of problems with your tests: 
1 . Memory is not tested
2. The naco_betapic.fits file is used, which is a toy sequence. It weights ~2MB whereas a normal cube is an order of magnitude larger. Maybe you start to imagine the problems we encounter when using 8 processes each one copying a 2GB array.

For the tests to be significant you'll need a real cube and to use the memory_profiler package (https://pypi.python.org/pypi/memory_profiler). It'll give you both the memory usage (with the -C option to include children processes) and running time. This might help: https://stackoverflow.com/questions/38358881/how-to-profile-multiple-subprocesses-using-python-multiprocessing-and-memory-pro

Reply:
From how I understand function calls and numpy arrays, the cube should never be copied. I'll look into the memory profiler to verify that, thank you for the links!

Reply:
And please include the results here for the posterity :)

Reply:
Shouldn't we move the entire discussion elsewhere? Issues, new PR?

Reply:
If the results show the global array brings no benefit then it will deserve a new PR definitively. For the moment we can just have the new test results here and decide. Even better, it would be great to create a notebook/gist and then link it here (easier to read and visualize).

Reply:
I tested the two variants with an already reduced SPHERE/IRDIS datacube of GJ504 (~410MB, 104×1024×1024) and `memory_profiler` — the results show a memory overhead for the function-argument-variant, which also varies in time while for the global-variant the memory usage stays constant (per annulus). Also, the inner/smaller rings are clearly faster with the global variant!
In the graphic, *top* is the global one and *bottom* where I pass the cube as function argument to the multiprocessing worker. The script was run with `nproc=20` on the `calc` server, with 25 annuli.

![benchmark result](https://user-images.githubusercontent.com/37106979/37819287-3a33ef8a-2e7d-11e8-81eb-0fc2bded32fb.png)